File: swift-prototype/Sources/BankingCore/Bank.swift
import Foundation

public class Account {
    public let id: String
    public let owner: String
    public private(set) var balance: Double

    public init(id: String, owner: String, initialBalance: Double) {
        precondition(initialBalance >= 0, "Saldo inicial não pode ser negativo")
        self.id = id
        self.owner = owner
        self.balance = initialBalance
    }

    fileprivate func debit(_ amount: Double) {
        balance -= amount
        // Erro irrecuperável: fatalError — equivalente ao panic! do Rust
        guard balance >= 0 else {
            fatalError("VIOLAÇÃO DE INVARIANTE: conta '\(id)' com saldo negativo (\(balance)). Estado corrompido.")
        }
    }

    fileprivate func credit(_ amount: Double) {
        balance += amount
    }
}

public class Bank {
    private var accounts: [String: Account] = [:]
    public private(set) var transactionLog: [String] = []

    public init() {}

    public func addAccount(_ account: Account) {
        accounts[account.id] = account
    }

    public func balance(of id: String) -> Double? {
        accounts[id]?.balance
    }

    // throws declara explicitamente que esta função pode falhar
    // O compilador exige try na chamada — análogo ao #[must_use] do Rust
    public func transfer(from fromId: String, to toId: String, amount: Double) throws -> Double {
        guard amount > 0 else {
            throw TransferError.negativeAmount(amount)
        }

        guard let fromAccount = accounts[fromId] else {
            throw TransferError.accountNotFound(id: fromId)
        }

        guard let toAccount = accounts[toId] else {
            throw TransferError.accountNotFound(id: toId)
        }

        guard amount <= fromAccount.balance else {
            throw TransferError.insufficientFunds(
                available: fromAccount.balance,
                requested: amount
            )
        }

        fromAccount.debit(amount)
        toAccount.credit(amount)

        let entry = String(format: "[OK] %@ → %@: R$%.2f (saldo restante: R$%.2f)",
                          fromId, toId, amount, fromAccount.balance)
        transactionLog.append(entry)

        return fromAccount.balance
    }

    // Retry com propagação de erro tipada
    public func transfer(
        from fromId: String,
        to toId: String,
        amount: Double,
        maxRetries: Int,
        simulateFailures: Int
    ) throws -> Double {
        for attempt in 0..<maxRetries {
            if attempt < simulateFailures {
                print("  [tentativa \(attempt + 1)/\(maxRetries)] Simulando timeout de rede...")
                Thread.sleep(forTimeInterval: 0.1)
                continue
            }
            return try transfer(from: fromId, to: toId, amount: amount)
        }
        throw TransferError.networkTimeout(retries: maxRetries)
    }
}

File: java-prototype/src/main/java/banking/Bank.java
package banking;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Bank {
    private final Map<String, Account> accounts = new HashMap<>();
    private final List<String> transactionLog = new ArrayList<>();

    public void addAccount(Account account) {
        accounts.put(account.id, account);
    }

    public Account getAccount(String id) {
        return accounts.get(id);
    }

    public List<String> getTransactionLog() {
        return List.copyOf(transactionLog);
    }

    // throws TransferException — o compilador exige que o chamador trate ou declare
    // Contraste com Rust: lá o compilador exige via #[must_use] no Result<T, E>
    public double transfer(String fromId, String toId, double amount)
            throws TransferException {

        if (amount <= 0) {
            throw new TransferException.NegativeAmount(amount);
        }

        Account from = accounts.get(fromId);
        if (from == null) {
            throw new TransferException.AccountNotFound(fromId);
        }

        Account to = accounts.get(toId);
        if (to == null) {
            throw new TransferException.AccountNotFound(toId);
        }

        if (amount > from.getBalance()) {
            throw new TransferException.InsufficientFunds(from.getBalance(), amount);
        }

        from.debit(amount);
        to.credit(amount);

        String entry = String.format("[OK] %s → %s: R$%.2f (saldo restante: R$%.2f)",
                fromId, toId, amount, from.getBalance());
        transactionLog.add(entry);

        return from.getBalance();
    }

    // ─────────────────────────────────────────────────────────────────
    // DEMONSTRAÇÃO CENTRAL DO SEMINÁRIO: o catch vazio
    //
    // Este método faz a mesma transferência que transfer(), mas captura
    // a exceção em um bloco vazio. O compilador Java ACEITA este código
    // sem qualquer aviso. A falha desaparece silenciosamente.
    //
    // Em Rust, o equivalente é IMPOSSÍVEL por design:
    //   bank.transfer("A", "B", 100.0);  // erro de compilação — Result ignorado
    // O atributo #[must_use] faz o compilador rejeitar código que descarta Result.
    // ─────────────────────────────────────────────────────────────────
    public void transferSemTratar(String fromId, String toId, double amount) {
        try {
            transfer(fromId, toId, amount);
        } catch (TransferException e) {
            // catch vazio — a falha é engolida, o chamador nunca saberá
            // o saldo permanece inalterado, mas nenhum sinal é emitido
        }
    }

    // Retry com re-lançamento tipado — mantém o tipo do erro original
    public double transferWithRetry(String fromId, String toId, double amount,
                                    int maxRetries, int simulateFailures)
            throws TransferException {

        for (int attempt = 0; attempt < maxRetries; attempt++) {
            if (attempt < simulateFailures) {
                System.out.printf("  [tentativa %d/%d] Simulando timeout de rede...%n",
                        attempt + 1, maxRetries);
                try {
                    Thread.sleep(100);
                } catch (InterruptedException ignored) {
                    Thread.currentThread().interrupt();
                }
                continue;
            }
            return transfer(fromId, toId, amount);
        }
        throw new TransferException.NetworkTimeout(maxRetries);
    }
}
